### ISSUE TITLE
Different results when using the same model from different sources 

### ISSUE THREAD

**Issue by Fcabla** [Babelscape/rebel#41]
Hello, I was thinking of using the REBEL model which is already trained to extract relationships from DBpedia abstracts.

Following the readme.md of this repository I could observe that the same model can be used from several sources, however when testing it I get different results:

Input text: "Barack Hussein Obama II is an American politician who is the 44th and current President of the United States. He is the first African American to hold the office and the first president born outside the continental United States. Born in Honolulu, Hawaii, Obama is a graduate of Columbia University and Harvard Law School, where he was president of the Harvard Law Review. He was a community organizer in Chicago before earning his law degree. He worked as a civil rights attorney and taught constitutional law at the University of Chicago Law School between 1992 and 2004. While serving three terms representing the 13th District in the Illinois Senate from 1997 to 2004, he ran unsuccessfully in the Democratic primary for the United States Hou"

When using the hugging faces model as you have described in the readme.md
```python
from transformers import pipeline
triplet_extractor = pipeline('text2text-generation', model='Babelscape/rebel-large', tokenizer='Babelscape/rebel-large')
...
```
I get the following relationships:

> Barack Hussein Obama II|educated at|Columbia University
> President of the United States|officeholder|Barack Hussein Obama II 
> Barack Hussein Obama II|position held|President of the United States
> Barack Hussein Obama II|educated at|Harvard Law School

Furthermore, when I use the model that I downloaded from the link in the readme.md and include it in the Spacy pipeline I get: 
```python
nlp = spacy.load("en_core_web_sm")
nlp.add_pipe("rebel", after="senter", config={
    'device':-1, 'model_name':'Babelscape/rebel-large'})
...
```
> community organizer|located in the administrative territorial entity|Chicago
> University of Chicago Law School|field of work|constitutional law
> Barack Hussein Obama II|position held|President of the United States        
> United States|ethnic group|African American
> Harvard Law Review|publisher|Harvard Law School
> President of the United States|officeholder|Barack Hussein Obama II
> Illinois Senate|country|United States Hou
> Democratic primary|country|United States Hou

I also tried with the code snippet that appears in the [hugging faces](https://huggingface.co/Babelscape/rebel-large) page, the one that appears at the bottom with the name 'Model and Tokenizer using transformers'. 
```python
...
# Load model and tokenizer
tokenizer = AutoTokenizer.from_pretrained("Babelscape/rebel-large")
model = AutoModelForSeq2SeqLM.from_pretrained("Babelscape/rebel-large")
gen_kwargs = {"max_length": 256,"length_penalty": 0,"num_beams": 3,"num_return_sequences": 3}
...
```
I got this results:
> Barack Hussein Obama|position held|President of the United States
> Barack Hussein Obama II|educated at|Harvard Law School
> Barack Hussein Obama II|educated at|Columbia University
> Barack Hussein Obama|educated at|Harvard Law School
> President of the United States|officeholder|Barack Hussein Obama
> Barack Hussein Obama II|place of birth|Honolulu, Hawaii
> Barack Hussein Obama|educated at|Columbia University
> Harvard Law Review|publisher|Harvard Law School
> Barack Hussein Obama II|position held|President of the United States
> President of the United States|officeholder|Barack Hussein Obama II

Maybe this is a silly question but, why do I get different behaviors if the model is the same? Am I doing something wrong? 

Do you know if applying some kind of preprocessing to the input text would improve the result of the model or does it work better without tweaking the input?

Thanks!

**Comment by LittlePea13** [comment]
Hi there, 

Well, the model should be the exact same. I think the changes are due to using different methods to generate the triplets. In the first case, you are using HF pipeline, second one it's the Spacy plugin and finally directly using the model with transformers generate method.

Seq2seq models are highly susceptible to changes in the decoding strategy/parameters. You can have different beams, use greedy decoding, incentivise longer sequences with the length penalty, etc. To have full control over all these, the last method (using transformers and the generate function) is the best way, however the other two are much easy to use, and if I am not mistaken you can also control for most of these parameters. Moreover, the Spacy plugin currently breaks down the text in sentences and inputs one at a time, which is an extra preprocessing step.

So, to sum up, these differences are expected, but you can play around with the parameters and input to obtain a decoding strategy close to your need.
